### PR TITLE
Feature referral login logic

### LINKED ISSUE #1000: Login authentication logic
On login button click:
- Pull OU attribute (pin)
- Pull OU ancestors and store the country level OU on a preference
- Compare provided pin with OU attribute (pin), and authorize when match

### PR DESCRIPTION
**Pull request by manuelplazaspalacio**
Close #1000

### CODE REVIEW

File: app/src/ereferrals/java/org/eyeseetea/malariacare/data/remote/PullOrganisationCredentials.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+import org.eyeseetea.malariacare.network.ServerAPIController;
+import org.hisp.dhis.client.sdk.models.attribute.AttributeValue;
+import org.hisp.dhis.client.sdk.models.organisationunit.OrganisationUnit;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.IOException;
+
+public class PullOrganisationCredentials {
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/data/remote/PullOrganisationCredentials.java]: @manuelplazaspalacio This intermediate class is necessary? I think that ServerAPIController should return dhis objects directly not json.

File: app/src/ereferrals/java/org/eyeseetea/malariacare/data/sync/importer/PullOrganisationCredentialsController.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+import android.content.Context;
+
+import org.eyeseetea.malariacare.data.IDataSourceCallback;
+import org.eyeseetea.malariacare.data.database.model.Program;
+import org.eyeseetea.malariacare.data.database.utils.PreferencesEReferral;
+import org.eyeseetea.malariacare.data.remote.PullOrganisationCredentials;
+import org.eyeseetea.malariacare.domain.entity.Credentials;
+
+public class PullOrganisationCredentialsController {
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/data/sync/importer/PullOrganisationCredentialsController.java]: This class is really a OrgUnitRepository and should return domain OrgUnits retrieving from API DHIS using SeverAPIController. Can you Rename and create a interface in domain for apply dependency inversion from UseCase where is invoked?

File: app/src/ereferrals/java/org/eyeseetea/malariacare/data/sync/importer/PullOrganisationCredentialsController.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                    }
+
+                    @Override
+                    public void onError(Throwable throwable) {
+                        callback.onError(throwable);
+                    }
+                });
+    }
+
+    public void pullUserProgram(final Callback callback) {
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/data/sync/importer/PullOrganisationCredentialsController.java]: if this class is OrgUnitRepository, pullUserProgram should be a method in a ProgramRepository class and return Program domain object. can you extract this to own class?

File: app/src/ereferrals/java/org/eyeseetea/malariacare/data/remote/PullOrganisationCredentials.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        if (!isNetworkAvailable) {
+            callback.onError(new NetworkException());
+        } else {
+            try {
+            JSONObject jsonObject = ServerAPIController.getOrganisationUnitsByCode(
+                    credentials.getUsername());
+                if (jsonObject == null) {
+                    callback.onError(new InvalidCredentialsException());
+                } else {
+                    Credentials OUCredentials = parseOrganisationUnitToCredentials(jsonObject);
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/data/remote/PullOrganisationCredentials.java]: the decision of convert OrgUnit data to credentials is a Domain decision not data decision. This decision should be realized second step in LoginUseCase after login with hardcoded credentials

File: app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/CheckCredentialsWithOrgUnitUseCase.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
+package org.eyeseetea.malariacare.domain.usecase;
+
+import org.eyeseetea.malariacare.data.database.utils.PreferencesEReferral;
+import org.eyeseetea.malariacare.domain.entity.Credentials;
+
+public class CheckCredentialsWithOrgUnitUseCase implements UseCase {
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/CheckCredentialsWithOrgUnitUseCase.java]: This really is not a domain use case, use case is only login but in erreferrals has two steps but this steps should be encapsulated in LoginUseCase

File: app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java [hunk 6]
```diff
@@ -18,7 +20,7 @@ public LoginUseCase(IAuthenticationManager authenticationManager) {
 
     @Override
     public void execute(Credentials credentials, final Callback callback) {
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java]: Login Use Case should has all login steps:
1 - Hardcoded login
2- Get OrgUnit y map to server credentials
3 - Verify user credentials with server credentials

File: app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java [hunk 7]
```diff
@@ -33,10 +35,21 @@ public void onError(Throwable throwable) {
                         } else if (throwable instanceof InvalidCredentialsException) {
                             callback.onInvalidCredentials();
                         } else if (throwable instanceof NetworkException) {
-                            callback.onNetworkError();
-                        } else {
-                            throwable.printStackTrace();
-                            callback.onUnexpectedError();
+                            CheckCredentialsWithOrgUnitUseCase checkCredentialsWithOrgUnitUseCase =
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java]: Why if an network error occurred then execute second step?

File: app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/PullOrganisationCredentialsUseCase.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
+package org.eyeseetea.malariacare.domain.usecase;
+
+import org.eyeseetea.malariacare.data.sync.importer.PullOrganisationCredentialsController;
+import org.eyeseetea.malariacare.domain.boundary.executors.IAsyncExecutor;
+import org.eyeseetea.malariacare.domain.boundary.executors.IMainExecutor;
+import org.eyeseetea.malariacare.domain.exception.InvalidCredentialsException;
+import org.eyeseetea.malariacare.domain.exception.NetworkException;
+import org.eyeseetea.malariacare.domain.exception.PullConversionException;
+
+public class PullOrganisationCredentialsUseCase implements UseCase {
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/PullOrganisationCredentialsUseCase.java]: This really is not a domain use case, use case is only login but in erreferrals has two steps but this steps should be encapsulated in LoginUseCase

File: app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/strategies/PullUseCaseStrategies.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    @Override
+    public void onPullComplete() {
+        mPullUseCase.notifyComplete();
+    }
+
+    @Override
+    public void onOnNetworkError() {
+        if (PreferencesEReferral.getUserProgramId() != -1) {
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/strategies/PullUseCaseStrategies.java]: Why this logic? If a network error occurs, should not it be communicated to the presentation layer via callbaks?

File: app/src/ereferrals/java/org/eyeseetea/malariacare/strategies/LoginActivityStrategy.java [hunk 10]
```diff
@@ -134,7 +145,64 @@ public void initViews() {
         passwordHint.setHint(loginActivity.getResources().getText(R.string.login_pin));
     }
 
-    public boolean onOptionsItemSelected(MenuItem item) {
-        return false;
+    @Override
+    public void onLoginSuccess(final Credentials credentials) {
+        PullOrganisationCredentialsController pullOrganisationCredentialsController =
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/strategies/LoginActivityStrategy.java]: Second login step should be realized by LoginUseCase

File: app/src/main/java/org/eyeseetea/malariacare/network/ServerAPIController.java [hunk 11]
```diff
@@ -726,6 +724,47 @@ public static boolean isUserClosed(String userUid) {
         return closedDate.before(new Date());
     }
 
+    public static JSONObject getOrganisationUnitsByCode(String code)
```
Comment (xurxodev) [inline app/src/main/java/org/eyeseetea/malariacare/network/ServerAPIController.java]: this method should returns domain OrgUnit not a jsonObject

File: app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java [hunk 12]
```diff
@@ -33,10 +35,21 @@ public void onError(Throwable throwable) {
...
+                                    new CheckCredentialsWithOrgUnitUseCase();
+                            checkCredentialsWithOrgUnitUseCase.execute(
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java]: A Use Case should not invoke another use case, when this happens it should be a clue that something in the design is not right

**Comment by manuelplazaspalacio** [comment]
Changes by @xurxodev  requested done, closing this issue

File: app/src/main/java/org/eyeseetea/malariacare/network/ServerAPIController.java [hunk 13]
```diff
@@ -726,6 +727,94 @@ public static boolean isUserClosed(String userUid) {
         return closedDate.before(new Date());
     }
 
+    public static org.eyeseetea.malariacare.domain.entity.OrgUnit getOrganisationUnitsByCode(
```
Comment (xurxodev) [inline app/src/main/java/org/eyeseetea/malariacare/network/ServerAPIController.java]: @manuelplazaspalacio After @ifoche merge this PR https://github.com/EyeSeeTea/pictureapp/pull/1074 There is already a method getCurrentOrgUnit with similar logic, can you refactor for unify code?

File: app/src/main/java/org/eyeseetea/malariacare/domain/entity/OrgUnit.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
+package org.eyeseetea.malariacare.domain.entity;
+
+
+public class OrgUnit {
```
Comment (xurxodev) [inline app/src/main/java/org/eyeseetea/malariacare/domain/entity/OrgUnit.java]: @manuelplazaspalacio After @ifoche merge this PR https://github.com/EyeSeeTea/pictureapp/pull/1074 There is already a domain entity OrganisationUnit with similar logic, can you refactor for unify code?

File: app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java [hunk 15]
```diff
@@ -1,45 +1,169 @@
...
         mAuthenticationManager = authenticationManager;
+        mMainExecutor = mainExecutor;
+        mAsyncExecutor = asyncExecutor;
     }
 
     @Override
-    public void execute(Credentials credentials, final Callback callback) {
-        mAuthenticationManager.login(credentials,
+    public void execute(final Credentials credentials, final Callback callback) {
+        mCallback = callback;
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java]: @manuelplazaspalacio All steps of login process should be executed with AsyncExecutor first step included. Review PushUseCase for view a example of when execute in async mode

File: app/src/ereferrals/java/org/eyeseetea/malariacare/data/remote/OrgUnitDataSource.java [hunk 16]
```diff
@@ -0,0 +1,16 @@
+package org.eyeseetea.malariacare.data.remote;
+
+import org.eyeseetea.malariacare.domain.boundary.repositories.IOrgUnitRepository;
+import org.eyeseetea.malariacare.domain.entity.Credentials;
+import org.eyeseetea.malariacare.domain.entity.OrgUnit;
+import org.eyeseetea.malariacare.domain.exception.NetworkException;
+import org.eyeseetea.malariacare.domain.exception.PullConversionException;
+import org.eyeseetea.malariacare.network.ServerAPIController;
+
+public class OrgUnitDataSource implements IOrgUnitRepository {
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/data/remote/OrgUnitDataSource.java]: @manuelplazaspalacio After @ifoche merge this PR https://github.com/EyeSeeTea/pictureapp/pull/1074 There is already a OrganisationUnitDataSource with similar logic, can you refactor for unify code?

File: app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java [hunk 17]
```diff
@@ -1,45 +1,169 @@
...
+                                    credentialsLocalDataSource.getOrganisationCredentials(), false);
                         }
                     }
                 });
 
     }
+
+
+    private void pullOrganisationCredentials(Credentials credentials, final Callback callback) {
+        IOrgUnitRepository orgUnitDataSource = new OrgUnitDataSource();
```
Comment (xurxodev) [inline app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java]: @manuelplazaspalacio can you refactor this dependencies for inject it in constructor?

**Comment by manuelplazaspalacio** [comment]
Changes by @xurxodev requested done, closing this issue

**Comment by manuelplazaspalacio** [comment]
Changes request by @ifoche done

**Pull request merged**